File: python/arikernel/runtime/autoscope.py
"""AutoScope — deterministic keyword-based task-to-preset classifier."""

from __future__ import annotations

from typing import Any

from .spec_loader import get_autoscope_config


def classify_scope(task: str) -> dict[str, Any]:
    """Classify a task description into a preset.

    Returns:
        {"preset": str, "confidence": float, "scores": dict}
    """
    config = get_autoscope_config()
    rules = config["rules"]
    fallback = config.get("fallback", "safe-research")
    min_confidence = config.get("minConfidence", 0.3)

    lower = task.lower()
    scores: dict[str, float] = {}

    for rule in rules:
        score = sum(1 for kw in rule["keywords"] if kw in lower)
        scores[rule["preset"]] = score

    best_preset = fallback
    best_score = 0.0
    total_score = sum(scores.values())

    for rule in rules:
        score = scores[rule["preset"]]
        if score > best_score:
            best_score = score
            best_preset = rule["preset"]

    confidence = best_score / total_score if total_score > 0 else 0.0

    if best_score == 0 or confidence < min_confidence:
        return {
            "preset": fallback,
            "confidence": 0.0,
            "scores": scores,
        }

    return {
        "preset": best_preset,
        "confidence": confidence,
        "scores": scores,
    }

File: packages/runtime/src/autoscope.ts
import type { PresetId } from "@arikernel/core";

interface ScopeRule {
	preset: PresetId;
	keywords: string[];
	weight: number;
}

const SCOPE_RULES: ScopeRule[] = [
	{
		preset: "safe-research",
		weight: 1,
		keywords: [
			"search",
			"browse",
			"summarize",
			"summary",
			"webpage",
			"website",
			"url",
			"article",
			"research",
			"google",
			"fetch",
			"scrape",
			"crawl",
			"news",
			"lookup",
			"find online",
			"web",
			"internet",
			"link",
		],
	},
	{
		preset: "rag-reader",
		weight: 2,
		keywords: [
			"document",
			"pdf",
			"csv",
			"json",
			"analyze",
			"analysis",
			"parse",
			"extract",
			"read file",
			"corpus",
			"retrieval",
			"rag",
			"vector",
			"embedding",
			"index",
			"knowledge base",
			"ingest",
		],
	},
	{
		preset: "workspace-assistant",
		weight: 3,
		keywords: [
			"code",
			"refactor",
			"implement",
			"fix bug",
			"write code",
			"edit file",
			"create file",
			"repo",
			"repository",
			"git",
			"commit",
			"test",
			"build",
			"compile",
			"debug",
			"lint",
			"format",
			"workspace",
			"project",
		],
	},
	{
		preset: "automation-agent",
		weight: 4,
		keywords: [
			"automate",
			"automation",
			"workflow",
			"pipeline",
			"sync",
			"crm",
			"api",
			"webhook",
			"schedule",
			"trigger",
			"integrate",
			"integration",
			"send email",
			"notify",
			"database",
			"records",
			"batch",
			"process",
		],
	},
];

export interface ScopeResult {
	preset: PresetId;
	confidence: number;
	scores: Record<PresetId, number>;
}

export function classifyScope(task: string): ScopeResult {
	const lower = task.toLowerCase();
	const scores: Record<string, number> = {};

	for (const rule of SCOPE_RULES) {
		let score = 0;
		for (const keyword of rule.keywords) {
			if (lower.includes(keyword)) {
				score += 1;
			}
		}
		scores[rule.preset] = score;
	}

	let bestPreset: PresetId = "safe-research";
	let bestScore = 0;
	let totalScore = 0;

	for (const rule of SCOPE_RULES) {
		const score = scores[rule.preset];
		totalScore += score;
		if (score > bestScore) {
			bestScore = score;
			bestPreset = rule.preset;
		}
	}

	const confidence = totalScore > 0 ? bestScore / totalScore : 0;

	// If no keywords matched or confidence is too low, fall back to safe-research
	if (bestScore === 0 || confidence < 0.3) {
		return {
			preset: "safe-research",
			confidence: 0,
			scores: scores as Record<PresetId, number>,
		};
	}

	return {
		preset: bestPreset,
		confidence,
		scores: scores as Record<PresetId, number>,
	};
}
